### PR TITLE
#13118 Implement the generic step in http services 

### PR DESCRIPTION
**Pull request by agreggio-slingr**
Pull-request for issue https://github.com/slingr-stack/platform/issues/13118 created by @agreggio-slingr

**What does it do?**
Adds a generic flow step

**How to test it?**

1. Config a new http-service per example: 
Upgrade Policy = Latest
Base URL = https://api.publicapis.org
Apply and push changes

2. Use http-service and create an action with Action Type -> Flow
In the flow designer create a new generic flow with:
method = Get
base url = /entries
Apply and push changes

3.Go to runtime and lauch action

**Implementation notes**
NA

**Deployment notes**
NA

### CODE REVIEW

File: appService.json [hunk 1]
```diff
@@ -77,6 +77,9 @@
             "description": "Sends a OPTIONS request"
         }
     ],
+    "flowSteps":  [
+        "generic"
```
Comment (juaniorbegozo) [inline appService.json]: Is it correct to call it "generic"? Can we use "http" instead?

File: flowSteps/generic/step.json [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		{
+			"label": "Method",
+			"name": "method",
+			"type": "choice",
+			"description": "Select method",
+			"defaultValue": "get",
+			"required": "true",
+			"options": {
+				"possibleValues": [
+					{
```
Comment (juaniorbegozo) [inline flowSteps/generic/step.json]: Remove "sum" option

File: appService.json (hunk 1 above)
Comment (sebasmoyano) [reply appService.json]: I would name it `HTTP call`.

**Comment by sebasmoyano** [comment]
Are we going to apply this to the http endpoint in a different story?

**Comment by sebasmoyano** [comment]
Can this be tested on dev?

File: flowSteps/generic/step.json [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+			"required": "true",
+			"options": {
+				"allowContextSelector": "false"
+			}
+		},
+		{
+			"name": "Headers",
+			"label": "headers",
+			"description": "The headers are JSON format, per example: {\"key\"=\"ABC123\",\"client\"=\"456789\".",
+			"type": "text",
```
Comment (sebasmoyano) [inline flowSteps/generic/step.json]: We should have a way to define a `key-value` to define this.

File: flowSteps/generic/step.json [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+			"options": {
+				"allowContextSelector": "false"
+			}
+		},
+		{
+			"name": "Params",
+			"label": "params",
+			"description": "The params are JSON format, per example: {\"key\"=\"ABC123\",\"client\"=\"456789\".",
+			"type": "text",
```
Comment (sebasmoyano) [inline flowSteps/generic/step.json]: We should have a way to define a key-value to define this.

File: flowSteps/generic/step.json [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+			"options": {
+				"allowContextSelector": "false"
+			}
+		},
+		{
+			"label": "Body",
+			"name": "body",
+			"type": "text",
```
Comment (sebasmoyano) [inline flowSteps/generic/step.json]: We should have a JSON (code) editor here.

File: flowSteps/generic/step.json [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+			"required": "['post', 'put', 'patch'].includes(config.method)",
+			"visibility": "['post', 'put', 'patch'].includes(config.method)",
+			"description": "This is used to set body request",
+			"options": {
+				"allowContextSelector": "false"
+			}
+		},
+		{
+			"label": "Required call back",
```
Comment (sebasmoyano) [inline flowSteps/generic/step.json]: Please check spelling: `Requires callback`.

File: flowSteps/generic/step.json [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+			"name": "requiredCallBack",
+			"type": "boolean",
+			"description": "If true it will return call back",
+			"defaultValue": false,
+			"options": {
+				"allowContextSelector": "false"
+			}
+		},
+		{
+			"label": "CallbackData",
```
Comment (sebasmoyano) [inline flowSteps/generic/step.json]: Please check spelling: `Callback data`.

File: flowSteps/generic/step.json [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+			"options": {
+				"allowContextSelector": "false"
+			}
+		}
+	],
+	"outputs": [
+		{
+			"label": "Response",
+			"name": "response",
+			"type": "text",
```
Comment (sebasmoyano) [inline flowSteps/generic/step.json]: The type should be a generic type. Maybe `object`? It s based on the content type.

**Comment by sebasmoyano** [comment]
Are we going to support different content types for body and response at this point?

**Comment by sebasmoyano** [comment]
I would have a new field called `Settings` with the following properties:
Screenshot from 2022-12-20 13-52-10

This is based on Postman. As initial version we can support the ones that are already implemented.

**Comment by agreggio-slingr** [comment]
@sebasmoyano 
> Are we going to apply this to the http endpoint in a different story?
I can apply both in the same issue.

> Can this be tested on dev?
We can't, I need to create a new tag from master

> Are we going to support different content types for body and response at this point?
Not at the moment, can we do it in the next iteration?

> I would have a new field called `Settings` with the following properties: Screenshot from 2022-12-20 13-52-10
> 
> This is based on Postman. As initial version we can support the ones that are already implemented.
Not at the moment, can we do it in the next iteration?

**Comment by sebasmoyano** [comment]
@agreggio-slingr at this stage we need to support:
- Headers: add support for key-value config.
- Params: add support for key-value config.
- Body: be able to enter a `JSON` (in the next iterations we need to support different content types).
- New field `Settings`: at least we need to support some of the listed configuration. In the next iterations we can add more properties.

File: flowSteps/httpCall/step.json [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+			"required": "['post', 'put', 'patch'].includes(config.method)",
+			"visibility": "['post', 'put', 'patch'].includes(config.method)",
+			"description": "This is used to set body request",
+			"options": {
+				"allowContextSelector": "false"
+			}
+		},
+		{
+			"label": "Requires callback",
+			"name": "requiredCallBack",
```
Comment (sebasmoyano) [inline flowSteps/httpCall/step.json]: Align the name based on the label.

File: flowSteps/httpCall/step.json [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+			"description": "This is used to set callback data",
+			"visibility": "config.requiredCallBack",
+			"options": {
+				"allowContextSelector": "false"
+			}
+		},
+		{
+			"label": "Callbacks",
+			"name": "callbacks",
+			"type": "text",
```
Comment (sebasmoyano) [inline flowSteps/httpCall/step.json:169]: Should be multiplicity many?

File: flowSteps/httpCall/step.json [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+						"label": "Trace",
+						"name": "trace"
+					}
+				],
+				"allowContextSelector": "false"
+			}
+		},
+		{
+			"name": "baseUrl",
+			"label": "Base URL",
```
Comment (sebasmoyano) [inline flowSteps/httpCall/step.json]: Should it be just `URL`?

**Comment by sebasmoyano** [comment]
> @agreggio-slingr at this stage we need to support:
> 
> * Headers: add support for key-value config.
> * Params: add support for key-value config.
> * Body: be able to enter a `JSON` (in the next iterations we need to support different content types).
> * New field `Settings`: at least we need to support some of the listed configuration. In the next iterations we can add more properties.

Once we have this ready let me know so I can test it on dev.

File: flowSteps/httpCall/step.json (hunk 10 above)
Comment (agreggio-slingr) [reply flowSteps/httpCall/step.json:169]: Callback not support multiplicity many.
The following information provided in the documentation was used for [approach](https://platform-docs.slingr.io/app_development_model_endpoints.html#callbacks).

**Comment by agreggio-slingr** [comment]
> Are we going to support different content types for body and response at this point?

Not in this version, because the sdk does not support different content types for body and response.

**Comment by agreggio-slingr** [comment]
> > @agreggio-slingr at this stage we need to support:
> > 
> > * Headers: add support for key-value config.
> > * Params: add support for key-value config.
> > * Body: be able to enter a `JSON` (in the next iterations we need to support different content types).
> > * New field `Settings`: at least we need to support some of the listed configuration. In the next iterations we can add more properties.
> 
> Once we have this ready let me know so I can test it on dev.

**Comment by agreggio-slingr** [comment]
> I would have a new field called `Settings` with the following properties: Screenshot from 2022-12-20 13-52-10
> 
> This is based on Postman. As initial version we can support the ones that are already implemented.

At the moment the sdk only supports the following configurations
followRedirects
download
fullResponse

Screenshot from 2022-12-22 14-21-28

**Comment by agreggio-slingr** [comment]
> Can this be tested on dev?

Once the merge to master is done, i can create a tag with -dev which allows you to test on dev and ci.

**Comment by agreggio-slingr** [comment]
> Are we going to apply this to the http endpoint in a different story?

Yes, here's the [PR](https://github.com/slingr-stack/http-endpoint/pull/3)

**Comment by sebasmoyano** [comment]
> > I would have a new field called `Settings` with the following properties: Screenshot from 2022-12-20 13-52-10
> > This is based on Postman. As initial version we can support the ones that are already implemented.
> 
> At the moment the sdk only supports the following configurations followRedirects download fullResponse
> 
> Screenshot from 2022-12-22 14-21-28

`Settings` should be only a label not a boolean field.

Please explain better the fields:
- followRedirects
- download
- fullResponse

**Comment by juaniorbegozo** [comment]
#approve

**Comment by agreggio-slingr** [comment]
> > > I would have a new field called `Settings` with the following properties: Screenshot from 2022-12-20 13-52-10
> > > This is based on Postman. As initial version we can support the ones that are already implemented.
> > 
> > 
> > At the moment the sdk only supports the following configurations followRedirects download fullResponse
> > Screenshot from 2022-12-22 14-21-28
> 
> `Settings` should be only a label not a boolean field.
> 
> Please explain better the fields:
> 
> * followRedirects
> * download
> * fullResponse



Resolve conversation
Screenshot from 2022-12-22 18-05-07

**Comment by sebasmoyano** [comment]
@agreggio-slingr I'm going to merge this so we can test this on dev.

#approve

**Pull request merged**